python/tank/templatekey.py

```diff
@@ -256,6 +273,145 @@ def _as_string(self, value):
+        """
+        if isinstance(format_spec, basestring) is False:
+            raise TankError("format_spec for TimestampKey %s is not of type string: %s" %
+                            (name, format_spec.__class__.__name__))
+        self.format_spec = format_spec
+
+        if isinstance(utc_default, bool) is False:
+            raise TankError("utc_default for TimestampKey %s is not of type boolean: %s" %
+                            (name, utc_default.__class__.__name__))
+        self.utc_default = utc_default
+
+        super(TimestampKey, self).__init__(
+            name,
+            default=self.__get_current_time
+        )
+
+    def __get_current_time(self):
+        """
+        Returns the current time as a datetime.datetime instance.
+
+        Do not streamline the code so the __init__ method simply passesd the dt.datetime.now method,
+        we can't mock datetime.now since it's builtin and will make unit tests more complicated to
+        write.
+
+        :returns: If utc_default, a datetime object representing time in the UTC timezone. Otherwise,
+                  a datetime object representing time in the local timezone. In any case, the tzinfo
+                  member is None.
+        """
+        if self.utc_default:
+            return dt.datetime.utcnow()
+        else:
+            return dt.datetime.now()
+
+    def validate(self, value):
+        """
+        Test if a value is valid for this key.
+
+        :param value: Value to test.
+
+        :returns: Bool
+        """
+        if isinstance(value, basestring) or isinstance(value, unicode):
+            return self._validate_str(value)
+        else:
+            return self._validate_value(value)
+
+    def _validate_str(self, str_value):
+        """
+        Test if a string is valid for this key.
+
+        :param str_value: String to test.
+
+        :returns: Bool
+        """
+        try:
+            self._as_value(str_value)
+            return True
+        except ValueError, e:
+            self._last_error = "Invalid string: %s" % e.message
+            return False
+        except TypeError, e:
+            self._last_error = "Invalid type: %s" % e.message
+            return False
+
+    def _validate_value(self, value):
+        """
+        Test if a string is valid for this key.
+
+        :param str_value: String to test.
+
+        :returns: Bool
+        """
+        try:
+            self._as_string(value)
+        except TypeError, e:
+            self._last_error = "Invalid type: %s" % e.message
+            return False
+        except ValueError:
+            self._last_error = "Invalid value: %s" % e.message
+            return False
+        return True
+
+    def _as_string(self, value):
+        """
+        Converts a given value as string.
+
+        :param value: Can be any of:
+                          - int or float representing seconds since epoch in local time.
+                          - datetime.datetime
```

Comment:
I felt that since there were so many places where time can come from in Python (datetime.{date,time,datetime}.now(), time.gmtime(), time.localtime(), time.now()) that it would be nice to offer all the known formats. I obviously had to pick on format when converting to string so I picked what I thought was best and that I would use when using the TimestampKey. I think the lack of symmetry comes from the fact that other data types have only one way to represent them, while type can be expressed in many ways.